libnetdata/url/url.c

```diff
@@ -52,17 +52,73 @@ char *url_decode(char *str) {
     return url_decode_r(buf, str, size);
 }
 
+//decode %XX character or return 0 if cannot
+char url_percent_escape_decode(char *s) {
+    if(likely(s[1] && s[2]))
+        return from_hex(s[1]) << 4 | from_hex(s[2]);
+    return 0;
+}
+
+//this (utf8 string related) should be moved in separate file in future
+char url_utf8_get_byte_length(char c) {
+    char length = 0;
+    while(likely(c & 0x80)) {
+        length++;
+        c <<= 1;
+    }
+    //4 byte is max size for UTF-8 char
+    //10XX XXXX is not valid character -> check length == 1
+    if(length > 4 || length == 1)
+        return -1;
+    //0XXX XXXX means 1 byte
+    return (length) ? length : 1;
+}
+
+//decode % encoded UTF-8 characters and copy them to *d
+//return count of bytes written to *d
+char url_decode_multibyte_utf8(char *s, char *d, char* d_end) {
+    char first_byte = url_percent_escape_decode(s);
+
+    if(unlikely(!first_byte))
+        return 0;
+
+    char byte_length = url_utf8_get_byte_length(first_byte);
+
+    if(unlikely(byte_length <= 0 || d+byte_length >= d_end))
+        return 0;
+
+    char to_read = byte_length;
+    while(to_read > 0) {
+        *d++ = url_percent_escape_decode(s);
+        s+=3;
+        to_read--;
+    }
+
+    return byte_length;
+}
+
 char *url_decode_r(char *to, char *url, size_t size) {
     char *s = url,           // source
          *d = to,            // destination
          *e = &to[size - 1]; // destination end
```

Comment:
Why not name them `src`, `dst` (or even `source`, `destination`), etc. In that case you dont need those comments.

Reply:
Hi,

They were called s, d and e (END) in original code. So I kind of tried to imitate original naming.

Reply:
Opsie. It is from original code, nvm then :see_no_evil: (in my opinion it makes no sense to name vars poorly and use comments to *decode* names).

Reply:
Yep agree with you. But as I mentioned I just kind of looked at code around (new to netdata) and tried to keep same 'style' although I will check what are actual official coding style rules (if any) when I get few free minutes tomorrow :)